File: backend/app/services/ai_service.py
# app/services/local_deepseek_service.py - Local DeepSeek Integration
import os
import json
import torch
import logging
from typing import List, Dict, Optional, Any
from datetime import datetime
from dataclasses import dataclass
import asyncio
from concurrent.futures import ThreadPoolExecutor
import threading

# Import DeepSeek components
try:
    from transformers import AutoTokenizer
    from deepseek.generate import generate, ModelArgs, Transformer
    DEEPSEEK_AVAILABLE = True
except ImportError as e:
    print(f"DeepSeek dependencies not available: {e}")
    DEEPSEEK_AVAILABLE = False

logger = logging.getLogger(__name__)

@dataclass
class ChatMessage:
    role: str  # 'user' or 'assistant'
    content: str
    timestamp: Optional[str] = None
    project_id: Optional[str] = None

class LocalDeepSeekService:
    """Local DeepSeek AI Service using local model files"""
    
    def __init__(self):
        self.model = None
        self.tokenizer = None
        self.is_initialized = False
        self.is_loading = False
        self.model_lock = threading.Lock()
        self.executor = ThreadPoolExecutor(max_workers=1, thread_name_prefix="deepseek")
        
        # Configuration from environment
        self.model_path = os.getenv("DEEPSEEK_MODEL_PATH", "./deepseek")
        self.config_path = os.getenv("DEEPSEEK_CONFIG_PATH", "./deepseek/config.json")
        self.max_tokens = int(os.getenv("DEEPSEEK_MAX_TOKENS", "200"))
        self.temperature = float(os.getenv("DEEPSEEK_TEMPERATURE", "0.7"))
        
        # Initialize model in background
        if DEEPSEEK_AVAILABLE and os.path.exists(self.model_path):
            self._initialize_model()
    
    def _initialize_model(self):
        """Initialize DeepSeek model"""
        if self.is_loading or self.is_initialized:
            return
            
        self.is_loading = True
        try:
            logger.info("Initializing local DeepSeek model...")
            
            # Check if model files exist
            if not os.path.exists(self.config_path):
                logger.error(f"Config file not found: {self.config_path}")
                return
            
            # Load configuration
            with open(self.config_path, 'r') as f:
                config_dict = json.load(f)
            
            # Setup model arguments
            model_args = ModelArgs(**config_dict)
            
            # Set device and dtype
            torch.cuda.set_device(0 if torch.cuda.is_available() else "cpu")
            torch.set_default_dtype(torch.bfloat16)
            torch.set_num_threads(8)
            torch.manual_seed(965)
            
            # Initialize model
            device = "cuda" if torch.cuda.is_available() else "cpu"
            with torch.device(device):
                self.model = Transformer(model_args)
            
            # Load tokenizer
            self.tokenizer = AutoTokenizer.from_pretrained(self.model_path)
            
            self.is_initialized = True
            logger.info("✅ Local DeepSeek model initialized successfully!")
            
        except Exception as e:
            logger.error(f"❌ Failed to initialize DeepSeek model: {e}")
            self.is_initialized = False
        finally:
            self.is_loading = False
    
    async def generate_response(
        self,
        message: str,
        project_title: str = "Current Project",
        project_context: str = "",
        conversation_history: List[ChatMessage] = None,
        max_tokens: int = None,
        temperature: float = None
    ) -> str:
        """Generate AI response using local DeepSeek model"""
        
        if not self.is_initialized:
            return self._generate_fallback_response(message, project_title, project_context)
        
        try:
            # Prepare parameters
            max_tokens = max_tokens or self.max_tokens
            temperature = temperature or self.temperature
            
            # Build prompt with context
            prompt = self._build_prompt(message, project_title, project_context, conversation_history)
            
            # Generate response in thread pool to avoid blocking
            loop = asyncio.get_event_loop()
            response = await loop.run_in_executor(
                self.executor, 
                self._generate_sync, 
                prompt, 
                max_tokens, 
                temperature
            )
            
            return self._format_response(response)
            
        except Exception as e:
            logger.error(f"DeepSeek generation error: {e}")
            return self._generate_fallback_response(message, project_title, project_context)
    
    def _generate_sync(self, prompt: str, max_tokens: int, temperature: float) -> str:
        """Synchronous generation method for thread execution"""
        with self.model_lock:
            try:
                # Tokenize input
                input_ids = self.tokenizer.encode(prompt)
                input_tensor = torch.tensor([input_ids])
                
                # Generate response
                with torch.no_grad():
                    output_ids = generate(
                        self.model, 
                        input_tensor, 
                        max_tokens, 
                        temperature
                    )
                
                # Decode response
                response = self.tokenizer.decode(output_ids[0], skip_special_tokens=True)
                
                # Remove input prompt from response
                if response.startswith(prompt):
                    response = response[len(prompt):].strip()
                
                return response
                
            except Exception as e:
                logger.error(f"Sync generation error: {e}")
                raise
    
    def _build_prompt(
        self, 
        message: str, 
        project_title: str, 
        project_context: str,
        conversation_history: List[ChatMessage]
    ) -> str:
        """Build conversation prompt for DeepSeek"""
        
        # System instruction
        system_prompt = f"""You are an intelligent AI assistant helping with the project "{project_title}".

Project Context: {project_context if project_context else 'Software development project'}

Your role:
- Provide practical, actionable advice
- Help with planning, technical decisions, and problem-solving
- Give specific solutions based on project context
- Be concise and professional
- Focus on software development best practices

"""
        
        # Add conversation history
        conversation = ""
        if conversation_history:
            for msg in conversation_history[-5:]:  # Last 5 messages
                role = "Human" if msg.role == "user" else "Assistant"
                conversation += f"{role}: {msg.content}\n"
        
        # Build final prompt
        prompt = f"{system_prompt}{conversation}Human: {message}\nAssistant:"
        
        return prompt
    
    def _format_response(self, response: str) -> str:
        """Format and clean the response"""
        if not response:
            return "I'm here to help with your project. Could you please rephrase your question?"
        
        # Clean the response
        response = response.strip()
        
        # Remove any "Assistant:" prefix if present
        if response.startswith("Assistant:"):
            response = response[10:].strip()
        
        # Limit response length
        if len(response) > 800:
            sentences = response.split('. ')
            truncated = []
            length = 0
            
            for sentence in sentences:
                if length + len(sentence) > 750:
                    break
                truncated.append(sentence)
                length += len(sentence) + 2
            
            response = '. '.join(truncated)
            if not response.endswith('.'):
                response += '.'
        
        return response
    
    def _generate_fallback_response(
        self, 
        message: str, 
        project_title: str, 
        project_context: str
    ) -> str:
        """Generate fallback response when model is not available"""
        
        msg_lower = message.lower().strip()
        
        # Greeting responses
        if any(word in msg_lower for word in ['hi', 'hello', 'hey', 'start']):
            return f"""👋 **Welcome to {project_title}!**

I'm your AI project assistant. I can help with:
• Project planning and task breakdown
• Technical guidance and best practices
• Problem-solving and debugging
• Code review and optimization

How can I assist you today?"""

        # Planning responses
        if any(word in msg_lower for word in ['plan', 'planning', 'strategy']):
            return f"""📋 **Project Planning for {project_title}**

Let's structure your approach:

**Phase 1: Setup & Planning**
• Define requirements and scope
• Choose technology stack
• Set up development environment

**Phase 2: Core Development**
• Implement main features
• Regular testing and reviews
• Iterative development

**Phase 3: Testing & Deployment**
• Comprehensive testing
• Performance optimization
• Production deployment

Which area would you like to focus on?"""

        # Technical help
        if any(word in msg_lower for word in ['error', 'bug', 'issue', 'problem']):
            return f"""🔧 **Technical Debugging for {project_title}**

**Systematic Approach:**
• Check logs and error messages
• Identify recent changes
• Test with minimal examples
• Use debugging tools

**Common Solutions:**
• Verify configurations and dependencies
• Check data formats and API calls
• Review recent code changes
• Test in isolated environment

Can you share more details about the specific issue?"""

        # Default response
        return f"""🤖 **AI Assistant for {project_title}**

I understand you're asking about: "{message}"

**I can help with:**
• Project planning and organization
• Technical implementation guidance
• Problem-solving strategies
• Best practices and recommendations

**Project Context:** {project_context if project_context else 'Ready to assist with your development needs'}

What specific aspect would you like to explore?"""

# Enhanced AI Service with Local DeepSeek Integration
class EnhancedLocalAIService:
    """Enhanced AI Service that prioritizes local DeepSeek"""
    
    def __init__(self):
        self.local_deepseek = LocalDeepSeekService()
        self.fallback_mode = not self.local_deepseek.is_initialized
        
    async def generate_smart_response(
        self,
        message: str,
        project_title: str = "Current Project",
        project_context: str = "",
        conversation_history: List[ChatMessage] = None,
        max_tokens: int = 200,
        temperature: float = 0.7
    ) -> str:
        """Generate intelligent response using local DeepSeek or fallback"""
        
        if self.local_deepseek.is_initialized:
            return await self.local_deepseek.generate_response(
                message, project_title, project_context, 
                conversation_history, max_tokens, temperature
            )
        else:
            return self.local_deepseek._generate_fallback_response(
                message, project_title, project_context
            )
    
    @property
    def is_initialized(self) -> bool:
        """Check if the service is properly initialized"""
        return self.local_deepseek.is_initialized or True  # Always return True as fallback works

# Global service instance
_local_ai_service_instance = None

def get_local_ai_service() -> EnhancedLocalAIService:
    """Get the local AI service instance"""
    global _local_ai_service_instance
    if _local_ai_service_instance is None:
        _local_ai_service_instance = EnhancedLocalAIService()
    return _local_ai_service_instance

File: flutter/choveenapp/lib/core/services/ai_service.dart
// lib/core/services/ai_service.dart
import 'dart:convert';
import 'dart:math';

class AIService {
  // Singleton pattern
  static final AIService _instance = AIService._internal();
  factory AIService() => _instance;
  AIService._internal();

  bool _isInitialized = false;
  bool get isInitialized => _isInitialized;

  // Initialize AI service
  Future<void> initialize() async {
    try {
      _isInitialized = true;
      print('✅ AI Service initialized successfully');
    } catch (e) {
      _isInitialized = false;
      print('❌ AI Service initialization failed: $e');
    }
  }

  // ✅ Get Smart AI Response with context awareness
  Future<String> getSmartAIResponse({
    required String message,
    String? projectTitle,
    String? projectContext,
    List<String>? userSkills,
    String? conversationHistory,
  }) async {
    try {
      // Simulate AI processing delay
      await Future.delayed(const Duration(milliseconds: 1500));

      final msgLower = message.toLowerCase();
      final skills = userSkills ?? ['Programming', 'Development'];
      final project = projectTitle ?? 'Your Project';
      
      // Context-aware intelligent responses
      if (_containsGreeting(msgLower)) {
        return _generateGreetingResponse(project, skills);
      }
      
      if (_containsPlanningKeywords(msgLower)) {
        return _generatePlanningResponse(project, skills);
      }
      
      if (_containsTechnicalKeywords(msgLower)) {
        return _generateTechnicalResponse(project, skills);
      }
      
      if (_containsHelpKeywords(msgLower)) {
        return _generateHelpResponse(project);
      }
      
      if (_containsTeamKeywords(msgLower)) {
        return _generateTeamResponse(project);
      }
      
      if (_containsTestingKeywords(msgLower)) {
        return _generateTestingResponse(project, skills);
      }
      
      if (_containsDeploymentKeywords(msgLower)) {
        return _generateDeploymentResponse(project, skills);
      }
      
      // Default intelligent response
      return _generateDefaultResponse(project, message, skills);
      
    } catch (e) {
      print('❌ AI Service error: $e');
      return _generateFallbackResponse(message);
    }
  }

  // ✅ Generate Project Suggestions
  Future<List<Map<String, dynamic>>> generateProjectSuggestions({
    required List<String> userSkills,
    List<String>? interests,
    String difficulty = 'intermediate',
  }) async {
    try {
      await Future.delayed(const Duration(milliseconds: 1000));
      
      final suggestions = <Map<String, dynamic>>[];
      final random = Random();
      
      // Base project templates based on skills
      final projectTemplates = _getProjectTemplates(userSkills, difficulty);
      
      for (int i = 0; i < min(5, projectTemplates.length); i++) {
        final template = projectTemplates[i];
        suggestions.add({
          'id': 'ai_proj_${random.nextInt(10000)}',
          'title': template['title'],
          'description': template['description'],
          'required_skills': userSkills.take(3).toList(),
          'category': template['category'],
          'difficulty': difficulty,
          'estimated_duration': template['duration'],
          'match_score': (0.9 - (i * 0.1)).clamp(0.5, 1.0),
          'ai_generated': true,
        });
      }
      
      return suggestions;
      
    } catch (e) {
      print('❌ AI Service error generating suggestions: $e');
      return _getFallbackSuggestions(userSkills, difficulty);
    }
  }

  // ✅ Keyword Detection Methods
  bool _containsGreeting(String msg) {
    return msg.contains(RegExp(r'\b(hello|hi|hey|greetings|سڵاو|سلام)\b'));
  }

  bool _containsPlanningKeywords(String msg) {
    return msg.contains(RegExp(r'\b(plan|planning|organize|roadmap|strategy|schedule)\b'));
  }

  bool _containsTechnicalKeywords(String msg) {
    return msg.contains(RegExp(r'\b(code|coding|programming|development|technical|bug|error|implementation)\b'));
  }

  bool _containsHelpKeywords(String msg) {
    return msg.contains(RegExp(r'\b(help|stuck|problem|issue|challenge|difficult)\b'));
  }

  bool _containsTeamKeywords(String msg) {
    return msg.contains(RegExp(r'\b(team|collaborate|members|communication|meeting)\b'));
  }

  bool _containsTestingKeywords(String msg) {
    return msg.contains(RegExp(r'\b(test|testing|quality|bug|debugging|qa)\b'));
  }

  bool _containsDeploymentKeywords(String msg) {
    return msg.contains(RegExp(r'\b(deploy|deployment|production|launch|release)\b'));
  }

  // ✅ Response Generation Methods
  String _generateGreetingResponse(String project, List<String> skills) {
    final skillsText = skills.take(3).join(', ');
    return '''👋 **Hello! Welcome to your AI Project Assistant**

I'm excited to help you build **$project**! I can see you have great skills in **$skillsText**.

🎯 **Here's how I can assist you today:**

**Project Planning**
• Break down your project into manageable phases
• Create realistic timelines and milestones
• Suggest best development practices

**Technical Guidance**
• Code architecture recommendations
• Technology stack suggestions
• Problem-solving support

**Team Collaboration**
• Workflow optimization
• Communication strategies
• Task management tips

What aspect of your project would you like to focus on first?''';
  }

  String _generatePlanningResponse(String project, List<String> skills) {
    return '''📋 **Smart Project Planning for $project**

🚀 **Recommended Development Phases:**

**Phase 1: Foundation** (Week 1-2)
• Define project requirements clearly
• Set up development environment
• Create basic project structure
• Plan database schema (if needed)

**Phase 2: Core Development** (Week 3-6)
• Implement main features step by step
• Build user interface components
• Add authentication & user management
• Create API endpoints & data handling

**Phase 3: Enhancement** (Week 7-8)
• Add advanced features
• Improve user experience
• Performance optimization
• Error handling & validation

**Phase 4: Finalization** (Week 9-10)
• Comprehensive testing
• Bug fixes & improvements
• Documentation
• Deployment preparation

💡 **Given your ${skills.take(2).join(' & ')} skills, I recommend starting with the core functionality first. What specific feature would you like to tackle?**''';
  }

  String _generateTechnicalResponse(String project, List<String> skills) {
    return '''🔧 **Technical Guidance for $project**

**Architecture Recommendations:**
• Use clean, modular code structure
• Implement proper error handling
• Follow ${skills.contains('Flutter') ? 'Flutter' : skills.contains('React') ? 'React' : 'industry'} best practices
• Consider scalability from the start

**Code Quality Tips:**
• Write descriptive variable and function names
• Add comments for complex logic
• Use version control (Git) consistently
• Implement unit tests where possible

**Common Issues & Solutions:**
• **Performance**: Optimize heavy operations, use lazy loading
• **Security**: Validate all inputs, use secure authentication
• **Maintainability**: Keep functions small and focused
• **User Experience**: Add loading states and error messages

**Debugging Strategy:**
1. Reproduce the issue consistently
2. Check logs and error messages
3. Use breakpoints and debugging tools
4. Test fixes in isolation

What specific technical challenge are you facing right now?''';
  }

  String _generateHelpResponse(String project) {
    return '''🆘 **I'm here to help with $project!**

**Let's solve this together:**

**Step 1: Describe the Problem**
• What exactly isn't working?
• When does the issue occur?
• What error messages do you see?

**Step 2: Gather Information**
• What were you trying to accomplish?
• What steps led to this issue?
• Have you made recent changes?

**Step 3: Troubleshoot**
• Check for typos in code
• Verify file paths and imports
• Look at console/log output
• Test in isolation

**Step 4: Find Solutions**
• Search documentation
• Check Stack Overflow
• Try alternative approaches
• Ask for specific help

**Common Quick Fixes:**
• Restart development server
• Clear cache/storage
• Update dependencies
• Check network connectivity

Tell me more about what you're stuck on, and I'll provide specific guidance!''';
  }

  String _generateTeamResponse(String project) {
    return '''👥 **Team Collaboration for $project**

**Effective Team Communication:**

**Daily Coordination**
• 15-minute daily standup meetings
• Share what you completed yesterday
• Discuss today's goals
• Identify any blockers

**Task Management**
• Use project management tools (Trello, Asana)
• Break work into small, clear tasks
• Assign ownership and deadlines
• Track progress visually

**Code Collaboration**
• Use Git for version control
• Create feature branches
• Write clear commit messages
• Review each other's code

**Knowledge Sharing**
• Document decisions and processes
• Share useful resources
• Conduct mini-learning sessions
• Create a team wiki

**Conflict Resolution**
• Address issues early and directly
• Focus on solutions, not problems
• Listen actively to all perspectives
• Seek compromise when possible

What specific team challenge would you like help with?''';
  }

  String _generateTestingResponse(String project, List<String> skills) {
    return '''🧪 **Testing Strategy for $project**

**Testing Pyramid Approach:**

**Unit Tests (Foundation)**
• Test individual functions and components
• Quick to run and easy to debug
• Should cover core business logic
• Aim for 70-80% of your tests here

**Integration Tests (Middle)**
• Test how components work together
• Verify API endpoints and database operations
• Check user workflows
• 15-20% of your test suite

**End-to-End Tests (Top)**
• Test complete user journeys
• Verify critical business scenarios
• Use tools like Selenium or Cypress
• 5-10% of tests, focus on key features

**${skills.contains('Flutter') ? 'Flutter' : skills.contains('React') ? 'React' : 'Mobile'} Specific Testing:**
• Widget/Component testing
• UI interaction testing
• Performance testing
• Device compatibility testing

**Testing Best Practices:**
• Write tests before fixing bugs
• Keep tests simple and focused
• Use descriptive test names
• Mock external dependencies

What type of testing would you like to implement first?''';
  }

  String _generateDeploymentResponse(String project, List<String> skills) {
    return '''🚀 **Deployment Guide for $project**

**Pre-Deployment Checklist:**

**Code Quality**
• All tests passing ✅
• Code reviewed and approved ✅
• No console errors or warnings ✅
• Performance optimized ✅

**Security**
• Environment variables secured ✅
• API keys and secrets protected ✅
• Input validation implemented ✅
• HTTPS enabled ✅

**${skills.contains('Flutter') ? 'Flutter App' : skills.contains('React') ? 'Web App' : 'Application'} Deployment:**

**Flutter Mobile:**
• Build release APK/IPA
• Test on physical devices
• Upload to Play Store/App Store
• Configure app signing

**Web Application:**
• Build production bundle
• Configure CDN and caching
• Set up domain and SSL
• Deploy to hosting platform

**Backend API:**
• Set up production database
• Configure environment variables
• Deploy to cloud service
• Set up monitoring and logs

**Post-Deployment:**
• Monitor application performance
• Set up error tracking
• Collect user feedback
• Plan future updates

Which deployment platform are you considering?''';
  }

  String _generateDefaultResponse(String project, String message, List<String> skills) {
    return '''🤖 **AI Assistant for $project**

I understand you're asking about: **"$message"**

Based on your ${skills.join(', ')} skills, here are some relevant suggestions:

**Immediate Actions:**
• Break down your question into smaller parts
• Check documentation for specific APIs or frameworks
• Look for similar examples in your codebase
• Consider alternative approaches

**Resources to Explore:**
• Official documentation
• Community forums and discussions
• Video tutorials and courses
• Code examples and repositories

**Next Steps:**
• Try implementing a simple version first
• Test with sample data
• Iterate and improve gradually
• Document your learning process

💡 **For more specific help, try asking:**
• "How do I implement [specific feature]?"
• "What's the best way to handle [specific scenario]?"
• "I'm getting [specific error], how to fix it?"

What specific aspect would you like me to elaborate on?''';
  }

  String _generateFallbackResponse(String message) {
    return '''🔄 **Processing your request: "$message"**

I'm currently working on understanding your question better. While I process this, here are some general tips:

**Development Best Practices:**
• Start with small, working increments
• Test frequently as you build
• Keep your code organized and commented
• Don't hesitate to refactor when needed

**When Stuck:**
• Take a step back and review the bigger picture
• Try explaining the problem to someone else
• Look for similar solutions online
• Break complex problems into smaller parts

**Resources:**
• Documentation is your best friend
• Stack Overflow for specific issues
• GitHub for code examples
• YouTube for step-by-step tutorials

Please feel free to rephrase your question or ask about something more specific!''';
  }

  // ✅ Project Templates
  List<Map<String, dynamic>> _getProjectTemplates(List<String> skills, String difficulty) {
    final templates = <Map<String, dynamic>>[];
    
    if (skills.any((s) => s.toLowerCase().contains('flutter') || s.toLowerCase().contains('mobile'))) {
      templates.addAll([
        {
          'title': 'Personal Task Manager App',
          'description': 'A complete task management app with categories, reminders, and progress tracking',
          'category': 'Mobile Development',
          'duration': '4-6 weeks',
        },
        {
          'title': 'Social Media Dashboard',
          'description': 'Create a social media management app with posting, analytics, and user engagement features',
          'category': 'Social App',
          'duration': '6-8 weeks',
        },
      ]);
    }
    
    if (skills.any((s) => s.toLowerCase().contains('web') || s.toLowerCase().contains('react') || s.toLowerCase().contains('javascript'))) {
      templates.addAll([
        {
          'title': 'E-commerce Website',
          'description': 'Build a full-featured online store with payment integration and admin panel',
          'category': 'Web Development',
          'duration': '8-10 weeks',
        },
        {
          'title': 'Portfolio Website',
          'description': 'Create a professional portfolio showcasing your projects and skills',
          'category': 'Personal Branding',
          'duration': '2-3 weeks',
        },
      ]);
    }
    
    // Add general templates
    templates.addAll([
      {
        'title': 'Learning Management System',
        'description': 'Build a platform for online courses with video streaming and progress tracking',
        'category': 'Education Technology',
        'duration': '10-12 weeks',
      },
      {
        'title': 'Chat Application',
        'description': 'Real-time messaging app with group chats, file sharing, and notifications',
        'category': 'Communication',
        'duration': '6-8 weeks',
      },
    ]);
    
    return templates;
  }

  List<Map<String, dynamic>> _getFallbackSuggestions(List<String> skills, String difficulty) {
    return [
      {
        'id': 'fallback_1',
        'title': 'Skill-Based Project',
        'description': 'A project tailored to your ${skills.join(', ')} skills',
        'required_skills': skills.take(3).toList(),
        'category': 'Custom Development',
        'difficulty': difficulty,
        'estimated_duration': '4-6 weeks',
        'match_score': 0.8,
        'ai_generated': false,
      }
    ];
  }
}